youtube_dl/extractor/panopto.py

```diff
@@ -0,0 +1,277 @@
+                        'title': 'OBJECT',
+                    },
+                },
+                {
+                    'md5': 'a483b8116abbb04a7112a9a3ccc835ce',
+                    'info_dict': {
+                        'id': '7668d6b2-dc81-421d-9853-20653689e2e8',
+                        'ext': 'mp4',
+                        'title': 'DV',
+                    },
+                },
+            ],
+            'playlist_count': 2,
+        },
+    ]
+
+    @staticmethod
+    def _get_contribs_str(contribs):
+        """Returns a comma-delimited string of contributors."""
+        s = ''
+        for c in contribs:
+            display_name = c.get('DisplayName')
+            if display_name is not None:
+                s += '{0}, '.format(display_name)
+        return s[:-2] if len(contribs) else ''
+
+    def _real_extract(self, url):
+        """Extracts the video and stream information for the given Panopto hosted URL."""
+        video_id = self._match_id(url)
+        org = self._match_organization(url)
+
+        delivery_info = self._download_json(
+            'https://{0}.hosted.panopto.com/Panopto/Pages/Viewer/DeliveryInfo.aspx'.format(org),
+            video_id,
+            query={
+                'deliveryId': video_id,
+                'invocationId': '',
+                'isLiveNotes': 'false',
+                'refreshAuthCookie': 'true',
+                'isActiveBroadcast': 'false',
+                'isEditing': 'false',
+                'isKollectiveAgentInstalled': 'false',
+                'isEmbed': 'false',
+                'responseType': 'json',
+            }
+        )
+
+        if 'ErrorCode' in delivery_info:
+            self._downloader.report_warning("If the video you are trying to download requires you to sign-in, you will "
+                                            "need to provide a cookies file that allows the downloader to authenticate "
+                                            "with Panopto. If the error below is about unauthorized access, this is "
+                                            "most likely the issue.")
+            raise ExtractorError(
+                'API error: ({0}) {1}'.format(delivery_info.get('ErrorCode', '?'), delivery_info.get('ErrorMessage', '?'))
+            )
+
+        streams = []
+        for this_stream in delivery_info['Delivery']['Streams']:
+            new_stream = {
+                'id': this_stream['PublicID'],
+                'title': this_stream['Tag'],
+                'formats': [],
+            }
+            if 'StreamHttpUrl' in this_stream:
+                new_stream['formats'].append({
+                    'url': this_stream['StreamHttpUrl'],
+                })
+            if 'StreamUrl' in this_stream:
+                m3u8_formats = self._extract_m3u8_formats(this_stream['StreamUrl'], video_id, 'mp4')
+                self._sort_formats(m3u8_formats)
+                new_stream['formats'].extend(m3u8_formats)
```

Comment:
```suggestion
                get_stream_url_matcher = lambda mid : re.compile(r'^https?://.*?/.*?/.*?' + mid + r'\?InvocationID.*')
                if '_M3U8_MATCHER' not in self.__dict__:
                    self._M3U8_MATCHER = get_stream_url_matcher(r'\.hls/master\.m3u8')
                if '_MP4_MATCHER' not in self.__dict__:
                    self._MP4_MATCHER = get_stream_url_matcher(r'\.mp4')
                if self._M3U8_MATCHER.match(this_stream['StreamUrl']):
                    m3u8_formats = self._extract_m3u8_formats(this_stream['StreamUrl'], video_id, 'mp4')
                    self._sort_formats(m3u8_formats)
                    new_stream['formats'].extend(m3u8_formats)
                elif self._MP4_MATCHER.match(this_stream['StreamUrl']):
                    new_stream['formats'].append({'url': this_stream['StreamUrl']})
                else:
                    raise ExtractorError('Unexpected StreamUrl format')
```

In some cases, it looks like StreamUrl is a single mp4 file instead of m3u8 (when there is only a single video I assume?). I don't know how fragile this regex is, but it passes all of my tests.